### PR TITLE
Merging SceneObjectNode and SceneObject

### PR DESCRIPTION
**Pull request by Licini**
The relation between `SceneObjectNode` and `SceneObject` was too convoluted. This PR merge them into one by having `SceneObject` inherent from `TreeNode`. This simplifies a lot on operations to access parent/ children object or a tree that an object is attachted too. While the most-outer level user APIs stays untouched.  This should also make it easier from creating a `compas_model` object.

```python
from compas.scene import Scene
from compas.geometry import Box
from compas.geometry import Sphere
from compas.geometry import Cylinder
from compas.geometry import Frame
from compas.geometry import Sphere



box = Box.from_width_height_depth(1, 1, 1)
cylinder = Cylinder(1, 1, Frame.worldXY())
sphere = Sphere(1)

scene = Scene()
scene.clear()

obj = scene.add(box).add(cylinder).add(sphere)

scene.print_hierarchy()


print(obj.tree)
```
```
└── <TreeNode: ROOT>
    └── <GeometryObject: Box>
        └── <GeometryObject: Cylinder>
            └── <GeometryObject: Sphere>
<Tree with 4 nodes>
```

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [x] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by tomvanmele** [comment]
i don't think this helps with the model problem. from an item you can access the parent node and from the parent node you can reach the tree. but you still can't reach the scene from the tree.

perhaps it would make sense to make `Scene` a subclass from `Tree`? might feel more logical since a scene object is a tree node. the problem of reaching the scene is them immediately solved.

File: src/compas/scene/sceneobject.py [hunk 1]
```diff
@@ -170,10 +167,12 @@ def add(self, item, **kwargs):
...
             If the scene object does not have an associated scene node.
         """
-        if self.node:
-            return self.node.add_item(item, **kwargs)
+        if isinstance(item, SceneObject):
+            sceneobject = item
         else:
-            raise ValueError("Cannot add items to a scene object without a node.")
+            sceneobject = SceneObject(item, **kwargs)
+        super().add(sceneobject)
```
Comment (tomvanmele) [inline src/compas/scene/sceneobject.py]: doesn't work in IronPython...

File: src/compas/scene/scene.py [hunk 2]
```diff
@@ -261,8 +114,14 @@ def add(self, item, parent=None, **kwargs):
...
-        sceneobject = SceneObject(item, context=self.context, **kwargs)
-        self.tree.add_object(sceneobject, parent=parent)
+
+        parent = parent or self.tree.root
+
+        if isinstance(item, SceneObject):
+            sceneobject = item
+        else:
+            sceneobject = SceneObject(item, context=self.context, **kwargs)
+        self.tree.add(sceneobject, parent=parent)
```
Comment (tomvanmele) [inline src/compas/scene/scene.py]: before, the problem was here. this might work now for the model...

**Comment by tomvanmele** [comment]
i added a few changes that make the scene work for recursive addition of objects in, for example a model.
added the type annotations for debugging.
we can remove them again...

**Comment by tomvanmele** [comment]
i removed viewer detection and the viewer instance attribute on the scene. as explained, i think we should rather do the opposite...

**Comment by Licini** [comment]
> i don't think this helps with the model problem. from an item you can access the parent node and from the parent node you can reach the tree. but you still can't reach the scene from the tree.
> 
> perhaps it would make sense to make `Scene` a subclass from `Tree`? might feel more logical since a scene object is a tree node. the problem of reaching the scene is them immediately solved.

I also think making `Scene` inherent from `Tree` is a good idea. Especially when SceneObject is the node. Will give this a try now on this branch

**Comment by Licini** [comment]
@tomvanmele 

1. How about now? `Scene` is now inheriting directly from `Tree`.
2. Regarding documenting the kwargs. Should we also explicitly write all of them out then pass on to super init, so it can be suggested pylance? For example:
```python
def __init__(kwag_x=None, kwag_y=None):
    super(XXX, self).__init__(kwag_x=kwag_x, kwag_y=kwag_y)
```

**Comment by tomvanmele** [comment]
regarding`**kwargs`...

in my opinion this should only be used when additional named arguments can be expected that are not known to the current class and need to be passed on to the class ancestors. arguments that are expected and used in the current scope should always be listed, processed, and documented explicitly...

what i think we should most certainly NOT do is the following:

```python
def __init__(self, **kwargs):
    super().__init__(**kwargs)
    self.a = kwargs.get('a')
    self.b = kwargs.get('b')
```

i would also vote for using ONLY named arguments in `__init__` functions in the future, because passing of arguments to parent classes is complicated when there is a mixture of positional and named arguments...

File: src/compas/scene/sceneobject.py [hunk 3]
```diff
@@ -7,21 +7,40 @@
...
 from functools import reduce
 from operator import mul
 
 
-class SceneObject(object):
+# if we know the parameters
+# we should document them
+# and not just make everything kwargs
+# not only is this not very helpful for the user
+# it does not allow for proper propagation of parameters across the inheritance chain
```
Comment (tomvanmele) [inline src/compas/scene/sceneobject.py]: perhaps remove this?

File: tests/compas/scene/test_scene.py [hunk 4]
```diff
@@ -80,34 +80,34 @@ def test_sceneobject_auto_context_discovery(mocker):
...
+    #     context.ITEM_SCENEOBJECT["Viewer"] = {FakeItem: FakeViewerSceneObject}
+    #     context.ITEM_SCENEOBJECT["Plotter"] = {FakeItem: FakePlotterSceneObject}
 
-        item = FakeSubItem()
-        sceneobject = SceneObject(item)
+    #     item = FakeSubItem()
+    #     sceneobject = SceneObject(item)
 
-        assert isinstance(sceneobject, FakeViewerSceneObject)
+    #     assert isinstance(sceneobject, FakeViewerSceneObject)
```
Comment (tomvanmele) [inline tests/compas/scene/test_scene.py]: let's remove this, no?

File: tests/compas/scene/test_scene.py [hunk 5]
```diff
@@ -80,34 +80,34 @@ def test_sceneobject_auto_context_discovery(mocker):
...
 
     def test_sceneobject_auto_context_discovery_no_context(mocker):
-        mocker.patch("compas.scene.context.is_viewer_open", return_value=False)
+        # mocker.patch("compas.scene.context.is_viewer_open", return_value=False)
```
Comment (tomvanmele) [inline tests/compas/scene/test_scene.py]: same

File: src/compas/scene/sceneobject.py [hunk 6]
```diff
@@ -148,8 +189,26 @@ def contrastcolor(self):
...
+            "opacity": self.opacity,
+            "show": self.show,
+        }
+
+        if self.frame:
+            settings["frame"] = self.frame
+        if self.transformation:
+            settings["transformation"] = self.transformation
+
+        return settings
```
Comment (tomvanmele) [inline src/compas/scene/sceneobject.py]: perhaps the existence of this container should be documented properly

File: src/compas/scene/sceneobject.py [hunk 7]
```diff
@@ -7,21 +7,40 @@
...
-    **kwargs : dict
-        Additional keyword arguments for constructing SceneObject.
+    name
+    color
+    opacity
+    show
+    frame
+    transformation
+    context
+    **kwargs
```
Comment (tomvanmele) [inline src/compas/scene/sceneobject.py]: still need to fill these in...

**Comment by Licini** [comment]
> regarding`**kwargs`...
> 
> in my opinion this should only be used when additional named arguments can be expected that are not known to the current class and need to be passed on to the class ancestors. arguments that are expected and used in the current scope should always be listed, processed, and documented explicitly...
> 
> what i think we should most certainly NOT do is the following:
> 
> ```python
> def __init__(self, **kwargs):
>     super().__init__(**kwargs)
>     self.a = kwargs.get('a')
>     self.b = kwargs.get('b')
> ```
> 
> i would also vote for using ONLY named arguments in `__init__` functions in the future, because passing of arguments to parent classes is complicated when there is a mixture of positional and named arguments...

Sounds good, will make everything named argument in the scope of this PR and make a separate one after for rest of places. And will fix all the docs and remove the code commented out

**Comment by Licini** [comment]
@tomvanmele cleaned up, shall we merge?

**Comment by Licini** [comment]
@tomvanmele let's merge this?

**Pull request merged**